libs/shacl/src/lib/services/shacl-data-graph-helper.service.ts

```diff
@@ -0,0 +1,225 @@
+      const pS = this.shaclShapesGraphHelper.getPropertyShapeInNodeShape(nodeShape, propertyPath);
+      if (pS) return pS;
+    }
+
+    return undefined;
+  }
+
+  isPointingToExternalReference(propertyShapeUri: string): boolean {
+    const propertyShape = this.shapesGraph.nodes[propertyShapeUri] as ShPropertyShape;
+    const editor = getOneNode(propertyShape[HANAMI.editor], this.shapesGraph);
+
+    if (editor && this.jsonLdService.isOfTypeGuard<HanamiNodeEditor>(editor, HANAMI.NodeEditor)) {
+      const linkingStrategy = getAllValues<LinkingStrategy>(editor[HANAMI.linkingStrategy], this.shapesGraph);
+      return linkingStrategy.length === 1 && linkingStrategy[0] === HANAMI.ExternalReferenceLinkingStrategy;
+    }
+
+    return false;
+  }
+
+  hasForm(nodeUri: string): boolean {
+    return this.hasAtLeastOneNonExternalReference(nodeUri) ? this.getPropertiesPaths(nodeUri).length > 0 : false;
+  }
+
+  private hasAtLeastOneNonExternalReference(nodeUri: string): boolean {
+    const allLinkingStrategies = new Set<LinkingStrategy>();
+
+    for (const [propertyPath, subjects] of Object.entries(this.objectPredicateSubject[nodeUri] ?? {})) {
+      for (const subjectUri of subjects) {
+        const propertyShape = this.getPropertyShape(subjectUri, propertyPath);
+        if (!propertyShape) continue;
+
+        const editor = getOneNode(propertyShape[HANAMI.editor], this.shapesGraph);
+        let linkingStrategies = isOfType<HanamiNodeEditor>(editor, HANAMI.NodeEditor)
+          ? getAllValues<LinkingStrategy>(editor[HANAMI.linkingStrategy], this.shapesGraph)
+          : [DEFAULT_LINKING_STRATEGY];
+        if (!linkingStrategies.length) {
+          linkingStrategies = [DEFAULT_LINKING_STRATEGY];
+        }
+
+        linkingStrategies.forEach(lS => allLinkingStrategies.add(lS));
+      }
+    }
+
+    return allLinkingStrategies.size !== 1 || !allLinkingStrategies.has(HANAMI.ExternalReferenceLinkingStrategy);
+  }
+
+  private getNodeShapesForTargetObjectsOf(nodeUri: string): ShNodeShape[] {
+    if (!this.shaclShapesGraphHelper.hasTargetObjectsOf) return [];
+    if (this.targetObjectsOfShapesByNodeUri[nodeUri]) return this.targetObjectsOfShapesByNodeUri[nodeUri];
+
+    const propertyPaths = Object.keys(this.objectPredicateSubject[nodeUri] ?? {});
+
+    return (this.targetObjectsOfShapesByNodeUri[nodeUri] = propertyPaths.reduce((acc, propertyPath) => {
+      const nodeShapeUris = this.shaclShapesGraphHelper.nodeShapeUrisByTargetObjectsOf[propertyPath];
+      nodeShapeUris?.forEach(nodeShapeUri => {
+        acc.push(this.shapesGraph.nodes[nodeShapeUri]);
+      });
+      return acc;
+    }, [] as ShNodeShape[]));
+  }
+
+  private getNodeShapesForShNode(nodeUri: string): ShNodeShape[] {
+    if (!this.shaclShapesGraphHelper.hasShNode) return [];
+    if (this.shNodeByNodeUri[nodeUri]) return this.shNodeByNodeUri[nodeUri];
+
+    const expectedNodeKind = isBlankNodeUri(nodeUri) ? SH.BlankNode : SH.IRI;
```

Comment:
we can use `getConcreteNodeKindOfValue`

Reply:
`getConcreteNodeKindOfValue` only works with `JsonLdValue`, not strings ^^'